### ISSUE TITLE
[3pt] Chain selection execution time grows linearly with the number of blocks

### ISSUE THREAD

**Issue by aaronschutza** [Topl/Prosomo#10]
The performance of each stakeholder degrades as the number of blocks increases. LSMStore is inefficient for parsing through block headers and checking the consensus properties.  Stakeholders must compete for disk access.  There are a lot of ways this could be optimized.  The goal is to have no change in performance as the number of blocks increases.

**Comment by aaronschutza** [comment]
Very pertinent article about the different types of maps available in scala that was recently written, will be useful for designing caches for recently added key/value pairs that preserve the ordering of insertion:

https://alvinalexander.com/scala/how-to-choose-map-implementation-class-sorted-scala-cookbook

**Comment by aaronschutza** [comment]
Cache has been implemented for block storage and state history.  Works well but the performance issues are not entirely related to disk access.  I noticed that the cpu load and update method execution time spikes as soon as there are 2 journal files in block storage directory.  I have a hunch that the LSMStore background task manager is interfering with simulation execution.  This might be fixed by switching to manual storage management.  I will make this a separate issue.  This issue is now closed singe chain selection is fast and efficient now with the cache functionality.
